pkg/endpoint/bpf.go

```diff
@@ -1302,16 +1321,18 @@ func (e *Endpoint) addPolicyMapDelta() error {
 // PolicyMap is unable to be dumped, or any update operation to the map fails.
 // Must be called with e.mutex Lock()ed.
 func (e *Endpoint) syncPolicyMapWithDump() error {
-
-	if e.realizedPolicy.PolicyMapState == nil {
-		e.realizedPolicy.PolicyMapState = make(policy.MapState)
-	}
-
 	if e.policyMap == nil {
 		return fmt.Errorf("not syncing PolicyMap state for endpoint because PolicyMap is nil")
 	}
 
-	currentMapContents, err := e.policyMap.DumpKeysToSlice()
+	// Apply pending policy map changes first so that desired map is up-to-date before
+	// we diff the maps below.
+	_, err := e.applyPolicyMapChanges()
+	if err != nil {
+		return err
+	}
+
+	currentMap, err := e.dumpPolicyMapToMapState()
```

Comment:
There's a small difference that you might have noticed which is previously we were only dumping the keys, not keys + values. Do we have some benchmarks, specially for the mem-allocs, for these changes vs the previous changes?

Reply:
Using the values for comparison is required for correct operation, and as this dump is only done on a periodically run controller I would not be so concerned about the memory allocations that are done to store the values in the map. While we could code the deletes directly to the dump loop itself to avoid storing the dump results in a map, we do need to store the dump keys and values in a map to be able to compare with the desired mapstate. This comparison between the dumped values and desired values was missing before this change.